### PR TITLE
Make able to compile with Clang

### PR DESCRIPTION
**Pull request by adamemerson**

### CODE REVIEW

**Comment by cbodley** [comment]
@adamemerson are you able to compile lz4, or did you have to cmake with `-DWITH_LZ4=OFF`? i've been seeing http://tracker.ceph.com/issues/23856 with clang

**Comment by adamemerson** [comment]
@cbodley I had no problems with LZ4 and didn't do anything to disable it.

**Comment by cbodley** [comment]
okthx. fedora 28 must have the fixed lz4 package (1.8.1+)

File: src/pybind/rgw/setup.py [hunk 1]
```diff
@@ -157,7 +192,24 @@ def cythonize(x, **kwargs):
 
 flags = get_python_flags()
 
-setup(
+cythized = cythonize(
```
Comment (tchaikov) [inline src/pybind/rgw/setup.py]: why is this necessary? or put in other words, why shall we run cythonize twice?

File: src/pybind/cephfs/setup.py [hunk 2]
```diff
@@ -9,7 +9,32 @@
...
+            cmd_so = compiler.compiler_so
+            cmd_so = [c for c in cmd_so if not
+                      (c == '-mcet' or c.startswith('-fcf-protection'))]
+            compiler.compiler = cmd
+            compiler.set_executable('compiler', ' '.join(cmd))
+            compiler.compiler_so = cmd_so
+            compiler.set_executable('compiler_so', ' '.join(cmd_so))
+
+distutils.sysconfig.customize_compiler = monkey_with_compiler
+customize_compiler = monkey_with_compiler
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/pybind/cephfs/setup.py]: Because we import cusomtize compiler specifically, if we don't set both we still get the old one.

THOUGH I'll just change it so when we call customize compiler we do it fully qualified and eliminate the spare.

File: src/pybind/cephfs/setup.py [hunk 3]
```diff
@@ -9,7 +9,32 @@
 import textwrap
 from distutils.ccompiler import new_compiler
 from distutils.errors import CompileError, LinkError
+import distutils.sysconfig
 from distutils.sysconfig import customize_compiler
+unwrapped_customize = distutils.sysconfig.customize_compiler
+
+clang = False
+
+def monkey_with_compiler(compiler):
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply src/pybind/cephfs/setup.py]: Good idea.

File: src/pybind/rgw/setup.py (hunk 1 above)
Comment (adamemerson) [reply src/pybind/rgw/setup.py]: It's not. It was a leftover from when I was exploring and trying to figure out where the flags kept being reintroduced. I'll pull it out.

**Comment by adamemerson** [comment]
All right! Refactored.

**Comment by tchaikov** [comment]
@adamemerson did you push the updated branch?

File: src/pybind/cephfs/setup.py [hunk 4]
```diff
@@ -9,7 +9,30 @@
...
 from distutils.errors import CompileError, LinkError
-from distutils.sysconfig import customize_compiler
+import distutils.sysconfig
+
+unwrapped_customize = distutils.sysconfig.customize_compiler
+
+clang = False
+
+def filter_unsupported_flags(flags):
+    if (clang):
```
Comment (tchaikov) [inline src/pybind/cephfs/setup.py]: could you remove the parenthesis around `clang`?

File: src/pybind/cephfs/setup.py [hunk 5]
```diff
@@ -9,7 +9,30 @@
...
+        return [f for f in flags if not (f == '-mcet' or
+                                         f.startswith('-fcf-protection'))]
+    else:
+        return flags
+
+def monkey_with_compiler(compiler):
+    unwrapped_customize(compiler)
+    if (compiler.compiler_type == 'unix'):
```
Comment (tchaikov) [inline src/pybind/cephfs/setup.py]: ditto

File: src/pybind/rgw/setup.py [hunk 6]
```diff
@@ -157,7 +180,7 @@ def cythonize(x, **kwargs):
 
 flags = get_python_flags()
 
-setup(
+s = setup(
```
Comment (tchaikov) [inline src/pybind/rgw/setup.py]: what is `s` for?

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-05-08_04:19:34-rados-wip-kefu-testing-2018-05-07-2312-distro-basic-mira/

failed tests are tracked by 

- http://tracker.ceph.com/issues/24077
- http://tracker.ceph.com/issues/24076

**Comment by tchaikov** [comment]
@dillaman could you take a quick look at changes in rbd?
